### DISCUSSION TITLE
Represent actual diameter of particles in Pyvista

### DISCUSSION THREAD

**Discussion by voferreira** [pyvista/pyvista#1984] [Q&A]
I'd like to use pyvista to make a video of a DEM simulation, but I don't know how I could use the actual diameter of the particles to represent them in the plotter.

For now, my code is like
``` plotter.add_mesh(df, scalars = df['Velocity'], cmap = 'turbo', render_points_as_spheres = True, point_size = df['Diameter'][0]) ```

But it is not returning the actual diameter of particles to the size of the spheres, of course.
How can I find the scale factor between the ``` point_size ``` value and the actual diameter of my particles?

Thank you upfront for the support :)

**Comment by MatthewFlamm** [comment]
I would look into the glyph filter which can put a sphere with a physical size at each location.  It is unclear to me what type `df` is in your code example, otherwise I could give you an example.

**Comment by voferreira** [comment]
Hey @MatthewFlamm! Thanks for replying!

The df is a pyvista unstructuredGrid with 9 properties (N Arrays), one of them is the 'Diameter' the of particles (they're all equal in diameter), and another is the 'Velocity' of the particles. I also have access to the position of the particles. When I run my code it gives me the following image:
![particles](https://user-images.githubusercontent.com/65647639/147943519-8c55fc35-16ee-440f-af60-bbfa166b5d1f.png)

Hope it clarifies the question. Please let me know if you still need any further info.

**Reply by akaszynski** [reply]
Can we get the data for this? I can probably generate an example using [glyph](https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSet.glyph.html).

See:
https://docs.pyvista.org/examples/01-filter/glyphs.html

**Reply by voferreira** [reply]
Of corse!
There you go.
[result_particles.20000.0000.zip](https://github.com/pyvista/pyvista/files/7804441/result_particles.20000.0000.zip)

**Reply by akaszynski** [reply]
Getting the following error:
```
End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of result_particles.20000.0000.zip or
        result_particles.20000.0000.zip.zip, and cannot find result_particles.20000.0000.zip.ZIP, period.
```

**Reply by voferreira** [reply]
Oh, I'm sorry. Try it again, I compressed again and updated it
[result_particles.20000.0000.zip](https://github.com/pyvista/pyvista/files/7804554/result_particles.20000.0000.zip)

**Reply by akaszynski** [reply]
Here's how I'd do it:
```py
import pyvista

ds = pyvista.read('result_particles.20000.0000.vtu')

# create a sphere with a diameter of 1
sphere = pyvista.Sphere(theta_resolution=10, phi_resolution=10)

# create spheres scaled to the diameter of the array "Diameter"
ds_glyph = ds.glyph(scale='Diameter', geom=sphere)

# plot
ds_glyph.plot(color='w', smooth_shading=True)
```
![tmp](https://user-images.githubusercontent.com/11981631/147994288-265a2492-94ad-4cd1-89b3-0ce1cc98aa50.png)

These particles are now (as far as I can tell) the correct size, since we're using the `"Diameter"` array to scale the size of the original Sphere. This is effectively what `render_points_as_spheres` does.

I would love to have a copy of the datasets for visualization if you can upload them and wouldn't mind us generating an example with them.

**Reply by akaszynski** [reply]
Also, this dataset is really cool.

```py
ds_glyph.plot(scalars='Velocity', smooth_shading=True, background='w', scalar_bar_args={'color': 'k'}, cmap='bwr')
```

![tmp](https://user-images.githubusercontent.com/11981631/147994543-5f2f954c-5559-403b-b424-e14661535728.png)

**Comment by voferreira** [comment]
Sure! It is a pleasure.
Since they're a bit heavy, I needed to share them through a cloud link.
By the way, those were obtained using Lethe (https://github.com/lethe-cfd/lethe).
Please let me know if the link is working fine.

(https://1drv.ms/u/s!Aq4CDCZlUzEz6KYdsD2Ru2we98wdNg?e=LfKiiH)
________________________________
De: Alex Kaszynski ***@***.***>
Enviado: segunda-feira, 3 de janeiro de 2022 21:31
Para: pyvista/pyvista ***@***.***>
Cc: Victor Oliveira Ferreira ***@***.***>; Author ***@***.***>
Assunto: Re: [pyvista/pyvista] Represent actual diameter of particles in Pyvista (Discussion #1984)


Also, this dataset is really cool.

ds_glyph.plot(scalars='Velocity', smooth_shading=True, background='w', scalar_bar_args={'color': 'k'}, cmap='bwr')

[tmp]<https://user-images.githubusercontent.com/11981631/147994543-5f2f954c-5559-403b-b424-e14661535728.png>

**Reply by akaszynski** [reply]
Link works. Thanks for the data!

**Reply by voferreira** [reply]
My pleasure! Thank you very much for the support!

**Reply by akaszynski** [reply]
Created a movie of all the datasets

https://user-images.githubusercontent.com/11981631/147997766-5bd97e9b-00c5-4a7b-8038-8423bb4140de.mp4

<details>
  <summary>Code</summary>

```py
import numpy as np
import pyvista

# create a sphere with a diameter of 1
sphere = pyvista.Sphere(theta_resolution=5, phi_resolution=5)

def create_particles(filename):
    ds = pyvista.read(filename)

    # create spheres scaled to the diameter of the array "Diameter"
    ds_glyph = ds.glyph(scale='Diameter', geom=sphere)

    # compute normalized velocity
    ds_glyph['Velocity_Norm'] = np.linalg.norm(ds_glyph['Velocity'], axis=1)
    return ds_glyph

# add first step
ds_glyph = create_particles('result_particles.0000.0000.vtu')

pl = pyvista.Plotter()
pl.background_color = 'w'
pl.add_mesh(
    ds_glyph, scalars='Velocity_Norm', smooth_shading=True, rng=[0, 1E-1], cmap='jet',
    scalar_bar_args={'color': 'k', 'title': 'Normalized Velocity'},
)
pl.camera_position = [
    (-0.11240211783521928, -0.10337119380671915, -0.8022166189105187),
    (-0.2981888771219628, -0.0075243223509701695, 0.009132743476183355),
    (0.9750521799556361, 0.03637433295231902, 0.2189752366502563)
]  # obtained from pl.show(return_cpos=True)

pl.open_movie("/tmp/tmp.mp4", framerate=5, quality=7)
pl.show(auto_close=False, interactive=False)
pl.write_frame()

for ts in range(250, 10250, 250):
    ds_glyph = create_particles(f'result_particles.{ts:04d}.0000.vtu')
    pl.mesh.overwrite(ds_glyph)
    pl.render()
    pl.write_frame()

pl.close()
```
</details>

**Reply by akaszynski** [reply]
And a linear interpolation movie:

https://user-images.githubusercontent.com/11981631/148001041-734585eb-9249-462a-9b10-c211770af698.mp4



<details>
  <summary>Code</summary>

```py
# Linear interpolation
from tqdm import tqdm




import numpy as np
import pyvista

# create a sphere with a diameter of 1
sphere = pyvista.Sphere(theta_resolution=5, phi_resolution=5)

# add first step
ds_glyph = create_particles('result_particles.0000.0000.vtu')

pl = pyvista.Plotter(off_screen=True, window_size=[400, 800])
pl.background_color = 'w'
pl.add_mesh(
    ds_glyph, scalars='Velocity_Norm', smooth_shading=True, rng=[0, 1E-1], cmap='jet',
    scalar_bar_args={'color': 'k', 'title': 'Normalized Velocity'},
)
pl.camera_position = [
    (-0.11240211783521928, -0.10337119380671915, -0.8022166189105187),
    (-0.2981888771219628, -0.0075243223509701695, 0.009132743476183355),
    (0.9750521799556361, 0.03637433295231902, 0.2189752366502563)
]  # obtained from pl.show(return_cpos=True)
pl.open_movie("/tmp/tmp.mp4", framerate=20, quality=5)
# pl.show(auto_close=False, interactive=False)
# pl.write_frame()


def basic_interpolation(now, nxt, frac):
    # generate a linear interpolation between now and the next step
    # must sort based on IDs
    now_id = np.argsort(now['ID'])
    nxt_id = np.argsort(nxt['ID'])
    interp = now.copy()

    vel = now['Velocity_Norm'][now_id]*(1 - frac) + nxt['Velocity_Norm'][nxt_id]*frac
    # breakpoint()
    interp.points = now.points[now_id]*(1 - frac) + nxt.points[nxt_id]*frac
    interp['Velocity_Norm'] = vel

    # finally, generate particles
    ds_glyph_interp = interp.glyph(scale='Diameter', geom=sphere)
    ds_glyph_interp.active_scalars_name = 'Velocity_Norm'
    return ds_glyph_interp


n_substeps = 16
steps = range(6000, 10250, 250)
n_total = (len(steps) - 1)*n_substeps
pbar = tqdm(total=n_total, desc="Writing Frames")
for step in range(len(steps)):
    if step == len(steps) - 1:
        break
    this_step = steps[step]
    next_step = steps[step + 1]
    ds_glyph_now = pyvista.read(f'result_particles.{this_step:04d}.0000.vtu')
    ds_glyph_now['Velocity_Norm'] = np.linalg.norm(ds_glyph_now['Velocity'], axis=1)
    ds_glyph_next = pyvista.read(f'result_particles.{next_step:04d}.0000.vtu')
    ds_glyph_next['Velocity_Norm'] = np.linalg.norm(ds_glyph_next['Velocity'], axis=1)

    for substep in range(n_substeps):
        interp = basic_interpolation(ds_glyph_now, ds_glyph_next, substep/n_substeps)
        pl.mesh.overwrite(interp)
        pl.write_frame()
        pbar.update()

pl.close()

```

</details>

**Reply by voferreira** [reply]
This is nice! Thank you for sharing!

**Reply by akaszynski** [reply]
@voferreira, I'd like to use this dataset as an example, but I forgot to save it locally. Would it be possible to upload it again?

**Reply by voferreira** [reply]
Sure! It is available here: https://1drv.ms/u/s!Aq4CDCZlUzEz6NxMHxMIntKWh4WcyQ?e=SNTDgb. If anything, feel free to reach out any time! I have been using pyvista extensively and it is a pleasure to collaborate :)

**Reply by akaszynski** [reply]
> Sure! It is available here: https://1drv.ms/u/s!Aq4CDCZlUzEz6NxMHxMIntKWh4WcyQ?e=SNTDgb. If anything, feel free to reach out any time! I have been using pyvista extensively and it is a pleasure to collaborate :)

Awesome! If you have any other datasets or examples you'd be comfortable me posting on PyVista's examples, please let me know. Also, we're always excited to add to our [External Examples](https://docs.pyvista.org/getting-started/external_examples.html), do you mind if I link https://github.com/voferreira/Post-processing?

**Reply by voferreira** [reply]
No, it would be cool. I'll prepare another dataset of the fluid side of this simulation and send it to you. It would be interesting to see both in the same example.

**Reply by akaszynski** [reply]
Email me at akascap@gmail.com or (perhaps better) or send me (or the group) a message at [PyVista Slack](http://slack.pyvista.org/).

**Reply by voferreira** [reply]
Sure! Do you mind adding my other e-mail on slack instead? It is vferreira@estudante.ufscar.br

**Reply by akaszynski** [reply]
Sure!

**Comment by bluetyson** [comment]
Yes, these are cool.
